sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java

```diff
@@ -62,310 +58,272 @@
-      private transient FluentBackoff retryBackoff; // defaults to no retries
-      private static final Logger LOG = LoggerFactory.getLogger(SnsWriterFn.class);
-      private static final Counter SNS_WRITE_FAILURES =
-          Metrics.counter(SnsWriterFn.class, "SNS_Write_Failures");
+      private final Write<T> spec;
+      private transient SnsAsyncClient client;
+      private final ConcurrentLinkedQueue<SnsWriteException> failures;
+      private final ConcurrentLinkedQueue<KV<WrappedSnsResponse<T>, WindowedValue<T>>> results;
 
-      private final Write spec;
-      private transient SnsClient producer;
-
-      SnsWriterFn(Write spec) {
+      SnsAsyncWriterFn(Write<T> spec) {
         this.spec = spec;
+        this.failures = new ConcurrentLinkedQueue<>();
+        this.results = new ConcurrentLinkedQueue<>();
       }
 
       @Setup
-      public void setup() throws Exception {
-        // Initialize SnsPublisher
-        producer = spec.getSnsClientProvider().getSnsClient();
-
-        retryBackoff =
-            FluentBackoff.DEFAULT
-                .withMaxRetries(0) // default to no retrying
-                .withInitialBackoff(RETRY_INITIAL_BACKOFF);
-        if (spec.getRetryConfiguration() != null) {
-          retryBackoff =
-              retryBackoff
-                  .withMaxRetries(spec.getRetryConfiguration().getMaxAttempts() - 1)
-                  .withMaxCumulativeBackoff(spec.getRetryConfiguration().getMaxDuration());
-        }
+      public void setup() {
+        this.client = spec.getSnsClientProvider().getSnsAsyncClient();
       }
 
+      @SuppressWarnings("FutureReturnValueIgnored")
       @ProcessElement
-      public void processElement(ProcessContext context) throws Exception {
-        PublishRequest request =
-            (PublishRequest) spec.getPublishRequestFn().apply(context.element());
-        Sleeper sleeper = Sleeper.DEFAULT;
-        BackOff backoff = retryBackoff.backoff();
-        int attempt = 0;
-        while (true) {
-          attempt++;
-          try {
-            PublishResponse pr = producer.publish(request);
-            context.output(pr);
-            break;
-          } catch (Exception ex) {
-            // Fail right away if there is no retry configuration
-            if (spec.getRetryConfiguration() == null
-                || !spec.getRetryConfiguration().getRetryPredicate().test(ex)) {
-              SNS_WRITE_FAILURES.inc();
-              LOG.info("Unable to publish message {} due to {} ", request.message(), ex);
-              throw new IOException("Error writing to SNS (no attempt made to retry)", ex);
-            }
+      public void processElement(ProcessContext context, BoundedWindow elementWindow)
+          throws IOException {
+        T element = context.element();
+        Instant timestamp = context.timestamp();
+        PaneInfo paneInfo = context.pane();
+
+        WindowedValue<T> windowedValue =
```

Comment:
Why do we need to create `WindowedValue` here if only `windowedValue.getValue()` used down there?